### PR TITLE
Feature/mango indexing

### PR DESCRIPTION
**Pull request by ErickFabian**
Adds public api to create indexes for querys and adds check for methods `find_by` and `where` to verify an index exists for the given query properties
## Example

```
User.find_by(username: 'username')
=> Dolly::IndexNotFoundError (Dolly::IndexNotFoundError)

Dolly::MangoIndex.create('username-json-index', ['username'])
=> {:result=>"created", :id=>"_design/4fd61cc70e6ddc6a4f416938008165c5609f49e9", :name=>"username-json-index"}

User.find_by(username: 'username')
# index takes around 4 minutes to build on the av database with 2575233 documents you can see the progress in http://127.0.0.1:5984/_utils/#/activetasks
=> #<User:0x00007fb5fbe0ac00 @doc={:_id=>"user/+04diego@gmail.com", :username=>"ErickFabian", :email=>"+04diego@gmail.com", :first_name=>"Diego", :last_name=>"Torres"}, @username="ErickFabian", @email="+04diego@gmail.com", @first_name="Diego", @last_name="Torres">

```
## Public Api

```
Dolly::MangoIndex.create(index_name: String, fields: Array, type: String(default: 'json'))
=> Creates an index given for the fields passed

Dolly::MangoIndex.all
=> Returns an array of all the indexes

Dolly::MangoIndex.find_by_fields(fields: Array(symbols))
=> Tries to find an index for the given array of symbols

Dolly::MangoIndex.delete_all
=> Deletes all the indexes created but skips the special index for id's that is created by default

Dolly::MangoIndex.delete(index_doc)
=> Deletes the given index document
```

### CODE REVIEW

File: lib/dolly/mango_index.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  class MangoIndex
+    class << self
+      DESIGN = '_index'
+
+      def all
+        Dolly::Document.connection.get(DESIGN)[:indexes]
+      end
+
+      def create(name, fields, type = 'json')
+        Dolly::Document.connection.post(DESIGN, build_index_structure(name, fields, type))
```
Comment (javierg) [inline lib/dolly/mango_index.rb]: You can use delegate to have `connection.post` method available

`delegate :post, to: Dolly::Document.connection`

File: lib/dolly/mango_index.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      end
+
+      def find_by_fields(fields)
+        all.find do |index_doc|
+          index_doc.dig(:def, :fields).map(&:keys).flatten == fields
```
Comment (javierg) [inline lib/dolly/mango_index.rb:25]: dig can return `nil` is that ok?

File: lib/dolly/mango_index.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        end
+      end
+
+      def delete_all
```
Comment (javierg) [inline lib/dolly/mango_index.rb:29]: this method is not being used, why did you add it?

File: lib/tasks/db.rake [hunk 4]
```diff
@@ -59,5 +59,17 @@ namespace :db do
...
 
+  namespace :index do
+    desc "Creates indexes for mango querys located in db/indexes/*.json"
+    task create: :environment do
+      indexes_dir = Rails.root.join('db', 'indexes')
+      files = Dir.glob File.join(indexes_dir, '**', '*.json')
+
+      files.each do |file|
+        index_data = JSON.parse(File.read(file))
+        Dolly::MangoIndex.create(index_data['name'], index_data['fields'])
```
Comment (javierg) [inline lib/tasks/db.rake:72]: I wonder if the freeze string literals direction works with rake files 🤔

**Review by javierg — COMMENTED** [review]
Great job, I like the create index interface

File: lib/dolly/mango_index.rb (hunk 3 above)
Comment (ErickFabian) [reply lib/dolly/mango_index.rb:29]: Added it as completeness to the api, to handle all the possible indexes i was thinking of upon running the create task, delete all the indexes and creating them from scratch, but i'm exploring a way to use the `rev` to update indexes

File: lib/tasks/db.rake (hunk 4 above)
Comment (ErickFabian) [reply lib/tasks/db.rake:72]: it does work i think, i recall rubocop throws a warning in rake files to add it

File: lib/dolly/mango_index.rb (hunk 2 above)
Comment (ErickFabian) [reply lib/dolly/mango_index.rb:25]: Yeah it's fine, the fields wont match and `find` will not return that index document

File: lib/dolly/mango_index.rb (hunk 1 above)
Comment (ErickFabian) [reply lib/dolly/mango_index.rb]: @javierg  `delegate`  is part of active support and i'd rather not use it so i refactored the class to use `Forwardable`

File: lib/dolly/mango_index.rb (hunk 3 above)
Comment (javierg) [reply lib/dolly/mango_index.rb:29]: If this is the case, please add a test demonstrating the usage of this.

**Review by alejandrodevs — APPROVED** [review]
:shipit:

**Review by javierg — APPROVED** [review]
:shipit:
